### PR TITLE
Charlesmchen/mute threads

### PR DESCRIPTION
**Pull request by charlesmchen**
* I had to fully convert the conversation settings view to pure `OWSTableViewController`.  Previously, we were reusing the custom cell configuration, etc. from the storyboard, but this became too painful.
   * In converting the conversation settings view, some bugs were fixed.

PTAL @michaelkirk

### CODE REVIEW

File: Signal/Images.xcassets/ic_mute_thread.imageset/Contents.json [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+      "scale" : "1x"
+    },
+    {
+      "idiom" : "universal",
+      "filename" : "ic_mute_thread@2x.png",
+      "scale" : "2x"
+    },
+    {
+      "idiom" : "universal",
+      "filename" : "ic_mute_thread@3x.png",
```
Comment (charlesmchen) [inline Signal/Images.xcassets/ic_mute_thread.imageset/Contents.json]: I bought this icon from the Noun Project.

File: Signal/src/environment/NotificationsManager.m [hunk 2]
```diff
@@ -149,6 +149,10 @@ - (void)notifyUserForIncomingMessage:(TSIncomingMessage *)message
                             inThread:(TSThread *)thread
                      contactsManager:(id<ContactsManagerProtocol>)contactsManager
 {
+    if (thread.isMuted) {
```
Comment (charlesmchen) [inline Signal/src/environment/NotificationsManager.m]: I think muting is this easy.

File: Signal/src/ViewControllers/OWSTableViewController.m [hunk 3]
```diff
@@ -256,7 +269,37 @@ - (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPa
     if (item.customRowHeight) {
         return [item.customRowHeight floatValue];
     }
-    return [super tableView:tableView heightForRowAtIndexPath:indexPath];
+    return kOWSTable_DefaultCellHeight;
+}
+
+- (nullable UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)sectionIndex
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/OWSTableViewController.m]: I added support for custom headers and footers to `OWSTableViewController`

File: Signal/src/ViewControllers/OWSTableViewController.h [hunk 4]
```diff
@@ -62,9 +69,10 @@ typedef UITableViewCell *_Nonnull (^OWSTableCustomCellBlock)();
 
 #pragma mark -
 
-@interface OWSTableViewController : UITableViewController
+@interface OWSTableViewController : UIViewController
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/OWSTableViewController.h]: I converted `OWSTableViewController` to not be a `UITableViewController`.

File: Signal/src/ViewControllers/OWSConversationSettingsTableViewController.m [hunk 5]
```diff
@@ -501,6 +706,136 @@ - (IBAction)durationSliderDidChange:(UISlider *)slider
...
+    if (self.thread.isMuted) {
+        UIAlertAction *action = [UIAlertAction actionWithTitle:NSLocalizedString(@"CONVERSATION_SETTINGS_UNMUTE_ACTION",
+                                                                   @"Label for button to unmute a thread.")
+                                                         style:UIAlertActionStyleDestructive
+                                                       handler:^(UIAlertAction *_Nonnull ignore) {
+                                                           [weakSelf setThreadMutedUntilDate:nil];
+                                                       }];
+        [actionSheetController addAction:action];
+    } else {
+#ifdef DEBUG
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/OWSConversationSettingsTableViewController.m]: I added a mute duration just for debugging purposes.

**Comment by charlesmchen** [comment]
### Settings before muting.

screen shot 2017-04-18 at 12 39 25 pm

### The mute action sheet.

screen shot 2017-04-18 at 12 40 37 pm


### Settings after muting.

screen shot 2017-04-18 at 12 39 31 pm

### Muted thread has small icon in navigation bar.

screen shot 2017-04-18 at 12 39 35 pm

### Muted threads are _not_ decorated in the home view.

screen shot 2017-04-18 at 12 39 16 pm

File: Signal/src/ViewControllers/OWSTableViewController.m (hunk 3 above)
Comment (michaelkirk) [inline Signal/src/ViewControllers/OWSTableViewController.m]: This is great, thanks.

**Pull request merged**